nixos/modules/tasks/auto-upgrade.nix

```diff
@@ -91,24 +112,36 @@ let cfg = config.system.autoUpgrade; in
       path = with pkgs; [ coreutils gnutar xz.bin gzip gitMinimal config.nix.package.out ];
 
       script = let
-          nixos-rebuild = "${config.system.build.nixos-rebuild}/bin/nixos-rebuild";
-        in
+        nixos-rebuild = "${config.system.build.nixos-rebuild}/bin/nixos-rebuild";
+        date     = "${pkgs.coreutils}/bin/date";
+        readlink = "${pkgs.coreutils}/bin/readlink";
+        shutdown = "${pkgs.systemd}/bin/shutdown";
+        upgradeFlag = optional (cfg.channel == null) "--upgrade";
+      in
         if cfg.allowReboot then ''
-            ${nixos-rebuild} boot ${toString cfg.flags}
-            booted="$(readlink /run/booted-system/{initrd,kernel,kernel-modules})"
-            built="$(readlink /nix/var/nix/profiles/system/{initrd,kernel,kernel-modules})"
-            if [ "$booted" = "$built" ]; then
-              ${nixos-rebuild} switch ${toString cfg.flags}
-            else
-              /run/current-system/sw/bin/shutdown -r +1
-            fi
-          '' else ''
+          ${nixos-rebuild} boot ${toString (cfg.flags ++ upgradeFlag)}
+          booted="$(${readlink} /run/booted-system/{initrd,kernel,kernel-modules})"
+          built="$(${readlink} /nix/var/nix/profiles/system/{initrd,kernel,kernel-modules})"
+          ${optionalString (cfg.rebootWindow != null) ''current_time="$(${date} +%H:%M)"''}
+
+          if [ "''${booted}" = "''${built}" ]; then
             ${nixos-rebuild} switch ${toString cfg.flags}
-        '';
+          ${optionalString (cfg.rebootWindow != null) ''
+            elif [[ "''${current_time}" < "${cfg.rebootWindow.lower}" ]] || \
+                 [[ "''${current_time}" > "${cfg.rebootWindow.upper}" ]]; then
```

Comment:
Still wrong. Example is `lower = 12:00`, `upper = 6:00`, `current = 18:00` -> This is within the window, but the code doesn't do it then.

After a bit of thinking, I'd say the best way to correctly implement this is to: Determine if `lower > upper`, and if so, flip them around. Then check whether `lower < cur < upper`, and toggle this result if `lower > upper` from before. If the result is true, do the reboot, otherwise don't.